docs/proposals/202105256-cluster-class-and-managed-topologies.md

```diff
@@ -0,0 +1,507 @@
+- Currently, (for a cluster with 3 machine deployments) this is possible by updating these three different objects representing the sets of worker nodes in the pool. An easier user experience would be to update a single object to enable the scaling of multiple sets of worker nodes.
+
+### Implementation Details/Notes/Constraints
+
+The following section provides details about the introduction of new types and modifications to existing types to implement the ClusterClass functionality.
+If instead you are eager to see an example of ClusterClass and how the Cluster object will look like, you can jump to the Behaviour paragraph.
+
+#### New API types
+##### ClusterClass
+This CRD is a collection of templates that describe the managed topology for one or more clusters.
+```golang
+// +kubebuilder:object:root=true
+// +kubebuilder:resource:path=clusterclasses,shortName=cc,scope=Namespaced,categories=cluster-api
+// +kubebuilder:storageversion
+
+// ClusterClass is a template which can be used to create managed topologies.
+type ClusterClass struct {
+  metav1.TypeMeta   `json:",inline"`
+  metav1.ObjectMeta `json:"metadata,omitempty"`
+
+  Spec ClusterClassSpec `json:"spec,omitempty"`
+}
+
+// ClusterClassSpec describes the desired state of the ClusterClass.
+type ClusterClassSpec struct {
+  // ControlPlane is a reference to a local struct that holds the details
+  // for provisioning the Control Plane for the Cluster.
+  ControlPlane LocalObjectTemplate `json:"controlPlane,omitempty"`
+
+  // Worker describes the worker nodes for the cluster.
+  // It is a collection of node types which can be used to create
+  // the worker nodes of the cluster.
+  // +optional
+  Worker WorkerClass `json:"worker,omitempty"`
+
+  // Infrastructure is a reference to a provider-specific template that holds
+  // the details for provisioning infrastructure specific cluster
+  // for the underlying provider.
+  // The underlying provider is responsible for the implementation
+  // of the template to an infrastructure cluster.
+  Infrastructure LocalObjectTemplate `json:"infrastructure,omitempty"`
+}
+
+// WorkerClass is a collection of deployment classes which pronounce
+//
+type WorkerClass struct {
+  // Deployments is a list of machine deployment types that can be used to create
+  // a set of worker plane nodes.
+  Deployments []DeploymentClass `json:"deployments,omitempty"`
+}
+
+
+// DeploymentClass serves as a template to define a set of worker nodes of the cluster
+// provisioned using the `ClusterClass`.
+type DeploymentClass struct {
+  // Class denotes a type of worker node present in the cluster
+  Class string `json:"class,omitempty"`
+
+
+  // Template is a local struct containing a collection of templates for creation of
+  // MachineDeployment objects representing a set of worker nodes.
+  Template TemplateClass `json:"template"`
```

Comment:
I'd expect TemplateClass to have a naming reference to what's templating, e.g `MachineTemplateClass`

Reply:
How about `DeploymentTemplateClass` instead, since it is a collection of templates for the MachineDeployment?

Reply:
Make sense, though I'd be also in favour of not using the term Deployment here either since there's already a core API CR with that name, any reason why we are using deployment rather than machineDeployment? e.g
```
    type Worker struct {
      MachineDeployments []ManagedMachineDeployment `json:”machineDeployments,omitempty”`
    }
```

Or if we don't want to couple it to this particular scalable resource may be:

```
    type Worker struct {
      nodePool []ManagedNodePool `json:”machineDeployments,omitempty”`
    }
```
or something similar?

Reply:
Changed :+1: